File: src/swift/4_BehavioralPatterns/1_ChainOfResponsibility/MethodChain.swift
import Foundation

class Creature : CustomStringConvertible
{
  var name: String
  var attack: Int
  var defense: Int

  init(name: String, attack: Int, defense: Int)
  {
    self.name = name
    self.attack = attack
    self.defense = defense
  }

  var description : String
  {
    return "Name: \(name), A = \(attack), D = \(defense)"
  }
}

class CreatureModifier
{
  let creature: Creature
  var next: CreatureModifier?

  init(creature: Creature)
  {
    self.creature = creature
  }

  func add(_ cm: CreatureModifier)
  {
    if (next != nil) {
      next!.add(cm)
    } else {
      next = cm
    }
  }

  func handle() {
    next?.handle()
  }
}

class DoubleAttackModifier : CreatureModifier
{
  override func handle()
  {
    print("Doubling \(creature.name)'s attack")
    creature.attack *= 2
    super.handle()
  }
}

class IncreaseDefenseModifier : CreatureModifier
{
  override func handle()
  {
    print("Increasing \(creature.name)'s defense")
    creature.defense += 3;
    super.handle()
  }
}

class NoBonusesModifier : CreatureModifier
{
  override func handle()
  {
    // nothing here
    print("No bonuses for you!")
    // we don't call super.handle()
  }
}

func main()
{
  let goblin = Creature(name: "Goblin", attack: 2, defense: 2)
  print(goblin)

  let root = CreatureModifier(creature: goblin)

  // later
  root.add(NoBonusesModifier(creature: goblin))

  print("Let's double the goblin's attack")
  root.add(DoubleAttackModifier(creature: goblin))

  print("Let's increase goblin's defense")
  root.add(IncreaseDefenseModifier(creature: goblin))

  root.handle()
  print(goblin)
}

main()

File: src/go/4_BehavioralPatterns/1_ChainOfResponsibility/methodchain.go
package chainofresponsibility

import "fmt"

type Creature struct {
  Name string
  Attack, Defense int
}

func (c *Creature) String() string {
  return fmt.Sprintf("%s (%d/%d)",
    c.Name, c.Attack, c.Defense)
}

func NewCreature(name string, attack int, defense int) *Creature {
  return &Creature{Name: name, Attack: attack, Defense: defense}
}

type Modifier interface {
  Add(m Modifier)
  Handle()
}

type CreatureModifier struct {
  creature *Creature
  next Modifier // singly linked list
}

func (c *CreatureModifier) Add(m Modifier) {
  if c.next != nil {
    c.next.Add(m)
  } else { c.next = m }
}

func (c *CreatureModifier) Handle() {
  if c.next != nil {
    c.next.Handle()
  }
}

func NewCreatureModifier(creature *Creature) *CreatureModifier {
  return &CreatureModifier{creature: creature}
}

type DoubleAttackModifier struct {
  CreatureModifier
}

func NewDoubleAttackModifier(
  c *Creature) *DoubleAttackModifier {
  return &DoubleAttackModifier{CreatureModifier{
    creature: c }}
}

type IncreasedDefenseModifier struct {
  CreatureModifier
}

func NewIncreasedDefenseModifier(
  c *Creature) *IncreasedDefenseModifier {
  return &IncreasedDefenseModifier{CreatureModifier{
    creature: c }}
}

func (i *IncreasedDefenseModifier) Handle() {
  if i.creature.Attack <= 2 {
    fmt.Println("Increasing",
      i.creature.Name, "\b's defense")
    i.creature.Defense++
  }
  i.CreatureModifier.Handle()
}

func (d *DoubleAttackModifier) Handle() {
  fmt.Println("Doubling", d.creature.Name,
    "attack...")
  d.creature.Attack *= 2
  d.CreatureModifier.Handle()
}

type NoBonusesModifier struct {
  CreatureModifier
}

func NewNoBonusesModifier(
  c *Creature) *NoBonusesModifier {
  return &NoBonusesModifier{CreatureModifier{
    creature: c }}
}

func (n *NoBonusesModifier) Handle() {
  // nothing here!
}

func main() {
  goblin := NewCreature("Goblin", 1, 1)
  fmt.Println(goblin.String())

  root := NewCreatureModifier(goblin)

  //root.Add(NewNoBonusesModifier(goblin))

  root.Add(NewDoubleAttackModifier(goblin))
  root.Add(NewIncreasedDefenseModifier(goblin))
  root.Add(NewDoubleAttackModifier(goblin))



  // eventually process the entire chain
  root.Handle()
  fmt.Println(goblin.String())
}